File: smart-streaming/src/main/scala/com/smart/kafka/KafkaProducer_Test.scala
package com.smart.kafka

import java.text.SimpleDateFormat
import java.util.{Date, Properties}

import com.alibaba.fastjson.JSONObject
import org.apache.kafka.clients.producer.{KafkaProducer, ProducerConfig, ProducerRecord}

import scala.util.Random

/**
  * kafka producer
  * Created by fc.w on 2017/4/17.
  */
object KafkaProducer_Test extends App{

  private val project_id = Array("wkzf", "yfyk")
  private val os_plant = Array("ios", "android")
  private val random = new Random(100)
  def projectId(): String = project_id(random.nextInt(project_id.length))
  def osPlant(): String = os_plant(random.nextInt(os_plant.length))

  val properties = new Properties
  properties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "h5:9092,h6:9092,h7:9092")
  properties.put(ProducerConfig.CLIENT_ID_CONFIG, "MsgProducer")
  properties.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringSerializer")
  properties.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringSerializer")

  val producer = new KafkaProducer[String, String](properties)
  while(true) {
    // 创建json串
    val event = new JSONObject()
    event.put("ip", "10.0.18.100")
    event.put("user_agent", "agent_aaaa")

    // 往kafka发送数据
    producer.send(new ProducerRecord[String, String]("test_pv", event.toString))
    println("Message sent: " + event.toString   +"     date :"+ date )

    //每隔200ms发送一条数据
    Thread.sleep(10000)
  }


  def date(): String = {
    val date = new Date()
    val dateFormat: SimpleDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    dateFormat.format(date)
  }
}

File: smart-streaming/src/main/scala/com/smart/kafka/KafkaProducerTest.java
package com.smart.kafka;

import org.apache.kafka.clients.producer.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Properties;

/**
 * kafka producer
 * Created by fc.w on 2017/4/14.
 */
public class KafkaProducerTest extends Thread {

    private static Logger LOG = LoggerFactory.getLogger(KafkaProducerTest.class);

    private final KafkaProducer<String, String> producer;
    private final String topic;
    private final boolean isAsync;

    public KafkaProducerTest(String topic, boolean isAsync) {
        Properties properties = new Properties();
        properties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "h5:9092,h6:9092,h7:9092");
        properties.put(ProducerConfig.CLIENT_ID_CONFIG, "MsgProducer");
        properties.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringSerializer");
        properties.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringSerializer");

        this.producer = new KafkaProducer<String, String>(properties);
        this.topic = topic;
        this.isAsync = isAsync;
    }

    @Override
    public void run() {
        int msgNo = 0;

        while (true) {
            String msg = "Msg: " + msgNo;
            String key = msgNo + "";
            if (isAsync) {//异步
                System.out.println(key);
                producer.send(new ProducerRecord<String, String>(this.topic,msg));
//                producer.send(new ProducerRecord<String, String>(this.topic, key, msg));
            } else {//同步
                producer.send(new ProducerRecord<String, String>(this.topic, key, msg),
                        new KafkaProducerCallback(System.currentTimeMillis(), key, msg));
            }
            try {
                Thread.sleep(1000L);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * 发送消息后的回调函数
     */
    class KafkaProducerCallback implements Callback {

        private final long startTime;
        private final String key;
        private final String msg;

        public KafkaProducerCallback(long startTime, String key, String msg) {
            this.startTime = startTime;
            this.key = key;
            this.msg = msg;
        }

        public void onCompletion(RecordMetadata recordMetadata, Exception e) {
            long elapsedTime = System.currentTimeMillis() - startTime;
            if (recordMetadata != null) {
                System.out.println(msg + " be sended to partition no : " + recordMetadata.partition());
            }
        }
    }

    public static void main(String[] args) {
        new KafkaProducerTest("recommend_group_topic",true).start();
    }

}
